pkg/csi/cinder/openstack/openstack_volumes.go

```diff
@@ -282,7 +282,8 @@ func (os *OpenStack) DetachVolume(instanceID, volumeID string) error {
 		}
 	}
 
-	return fmt.Errorf("disk: %s has no attachments or not attached to compute %s", volume.ID, instanceID)
+	// Disk has no attachments or not attached to provided compute
+	return nil
```

Comment:
@axelgobletbdr  Thanks for the PR. Looks good. 
Just trying to understand better, in your case , If volume has no attachments , doesn't it mean that volume is in `Available` state  , then this function shouldn't throw error as per L264?

Reply:
@ramineni if the status is `Available` that is indeed correct. This does not necessarily need to be the case though. The volume can also already be attached to another instance. in that case, the state will be `in-use`, and the code will return an error at L285.

Reply:
@axelgobletbdr  right . I want to confirm that it fails only in case of multiattach volume type. Thanks

Reply:
It can happen for single attach volumes as well, through the following process:
1. Volume is attached to node X, and Kubernetes `VolumeAttachment` exists
2. Volume is somehow externally detached from node X. Kubernetes `VolumeAttachment` still exists
3. Pod using the volume is scheduled to a different node Y
4. Kubernetes will try to delete the current `VolumeAttachment` and create a new one. 
5. The deletion will not succeed due to the finalizers that need to be removed by the CSI plugin. The CSI plugin fails when the `ControllerUnpublishVolume` endpoint is called, since the volume is already scheduled to a different node.

It is quite an edge case. This also appeared for us because we are having some volume issues with our openstack deployment. Nevertheless, Kubernetes should be able to reconcile and end up in the expected state. It will be able to if the function call is idempotent as defined in the CSI spec.